rust/src/storage/retrieve.rs

```diff
@@ -1,230 +1,39 @@
-                NVTKey::Dependencies => {
-                    matches!(field, Field::NVT(NVTField::Dependencies(_)))
-                }
-                NVTKey::RequiredKeys => {
-                    matches!(field, Field::NVT(NVTField::RequiredKeys(_)))
-                }
-                NVTKey::MandatoryKeys => {
-                    matches!(field, Field::NVT(NVTField::MandatoryKeys(_)))
-                }
-                NVTKey::ExcludedKeys => {
-                    matches!(field, Field::NVT(NVTField::ExcludedKeys(_)))
-                }
-                NVTKey::RequiredPorts => {
-                    matches!(field, Field::NVT(NVTField::RequiredPorts(_)))
-                }
-                NVTKey::RequiredUdpPorts => {
-                    matches!(field, Field::NVT(NVTField::RequiredUdpPorts(_)))
-                }
-                NVTKey::Preference => {
-                    matches!(field, Field::NVT(NVTField::Preference(_)))
-                }
-                NVTKey::Reference => matches!(field, Field::NVT(NVTField::Reference(_))),
-                NVTKey::Category => matches!(field, Field::NVT(NVTField::Category(_))),
-                NVTKey::Family => matches!(field, Field::NVT(NVTField::Family(_))),
-                NVTKey::NoOp => matches!(field, Field::NVT(NVTField::NoOp)),
-                // TODO: in memory and file should map in this case
-                NVTKey::Nvt => matches!(field, Field::NVT(NVTField::Nvt(_))),
-            },
-
-            Retrieve::KB(s) => {
-                if let Field::KB(kb) = field {
-                    &kb.key == s
-                } else {
-                    false
-                }
-            }
-
-            Retrieve::NotusAdvisory(_) => matches!(field, Field::NotusAdvisory(_)),
-            Retrieve::Result(None) => matches!(field, Field::Result(_)),
-            Retrieve::Result(Some(id)) => {
-                if let Field::Result(r) = field {
-                    &r.id == id
-                } else {
-                    false
-                }
-            }
-        }
-    }
-}
+use super::error::StorageError;
 
-/// Result of a heap stored iterator or StorageError
-pub type FieldResult = Result<Box<dyn Iterator<Item = Field>>, StorageError>;
-
-fn retry<T, F>(f: F, max: u64) -> Result<T, StorageError>
-where
-    F: Fn() -> Result<T, StorageError>,
-{
-    if max == 0 {
-        return Err(StorageError::RetryExhausted);
-    }
-    let result = f();
-    if let Err(StorageError::Retry(reason)) = result {
-        tracing::debug!(reason, "retriever implementation returned retry error");
-        retry(f, max - 1)
-    } else {
-        result
-    }
-}
-
-/// Result of a heap stored iterator or StorageError
-pub type FieldKeyResult = Result<Box<dyn Iterator<Item = (ContextKey, Field)>>, StorageError>;
 /// Retrieves fields based on a key and scope.
-pub trait Retriever: Send + Sync {
+pub trait Retriever<KEY> {
```

Comment:
How do I poll data with this new construct? Given that I want don't want to retrieve all VTs at once, serialize that and send it out in a big chunk because if first byte limitation, I  usually want to get push and serialize smaller chunks depending on the speed of the client.

From a trait point of view this was previously possible as retrieve was an iterator of fields so a underlying implementation could handle that internally. I don't fully understand how I would achieve that use case with the new trait.

Reply:
Not sure what exactly do you mean, but it  seems, that this was also not done with the old implementation. If all or a part of NVTs was requested from a storage, all of them (also including Notus Advisories) were copied into a new struct and a iterator over that was returned. So internally I currently don't return a Iterator, but rather the new created struct to let the caller decide what to do with that. This was also only implemented for the `DefaultDispatcher` (Now `InMemoryStorage`). Maybe I miss, what you mean. In that case can you provide an example, what's not possible anymore?